src/mod_cache_users.erl

```diff
@@ -0,0 +1,258 @@
+stop_cache(HostType) ->
+    ok = ejabberd_sup:stop_child(tbl_name(HostType)).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+-record(cache_state, {tab :: atom(), ttl = timer:seconds(30) :: non_neg_integer()}).
+
+rotate_and_purge_last_segment(#cache_state{tab = ParentTab}) ->
+    First = ets:first(ParentTab),
+    Last = ets:last(ParentTab),
+    EtsSegment = case ets:info(ParentTab, size) of
+                     1 ->
+                         ets:lookup_element(ParentTab, First, 2);
+                     _ ->
+                         [{_, E}] = ets:take(ParentTab, First),
+                         true = ets:insert_new(ParentTab, {Last + 1, E}),
+                         E
+                 end,
+    ets:delete_all_objects(EtsSegment),
+    ok.
+
+del_member(_, _, '$end_of_table') ->
+    ok;
+del_member(ParentTab, Key, SegmentKey) ->
+    case ets:lookup(ParentTab, SegmentKey) of
+        [{_, EtsSegment}] ->
+            ets:delete(EtsSegment, Key),
+            del_member(ParentTab, Key, ets:next(ParentTab, SegmentKey));
+        [] ->
+            del_member(ParentTab, Key, ets:next(ParentTab, SegmentKey))
+    end.
+
+del_domain(_, _, '$end_of_table') ->
+    ok;
+del_domain(ParentTab, Domain, SegmentKey) ->
+    case ets:lookup(ParentTab, SegmentKey) of
+        [{_, EtsSegment}] ->
+            ets:match_delete(EtsSegment, {{Domain, '_'}}),
+            del_domain(ParentTab, Domain, ets:next(ParentTab, SegmentKey));
+        [] ->
+            del_domain(ParentTab, Domain, ets:next(ParentTab, SegmentKey))
+    end.
+
+
+-spec start_link(atom(), gen_mod:module_opts()) -> {ok, pid()}.
+start_link(ParentTab, Opts) ->
+    gen_server:start_link(?MODULE, [ParentTab, Opts], []).
+
+init([ParentName, Opts]) ->
+    pg2:create(ParentName),
+    pg2:join(ParentName, self()),
+    TTL = gen_mod:get_opt(ttl, Opts, infinity),
+    N   = gen_mod:get_opt(number_of_segments, Opts, 1),
+    EtsOpts = [ordered_set, named_table, protected, {read_concurrency, true}],
+    ets:new(ParentName, EtsOpts),
+    lists:foreach(
+      fun(I) ->
+              %% Note, these names are only for debugging purposes,
+              %% you're not supposed to use the name as an API
+              SegmentTab = list_to_atom(atom_to_list(ParentName) ++ "_" ++ integer_to_list(I)),
+              SegmentOpts = [set, named_table, public, {read_concurrency, true}, {write_concurrency, true}],
+              Ref = ets:new(SegmentTab, SegmentOpts),
+              ets:insert(ParentName, {I, Ref})
+      end, lists:seq(1, N)),
+    erlang:send_after(TTL, self(), purge),
+    {ok, #cache_state{tab = ParentName, ttl = TTL}}.
+
+handle_call(Msg, From, State) ->
+    ?UNEXPECTED_CALL(Msg, From),
+    {reply, ok, State}.
+
+handle_cast({remove_user, Key}, #cache_state{tab = ParentTab} = State) ->
+    del_member(ParentTab, Key, ets:first(ParentTab)),
```

Comment:
we could wrap handle_remove_user and handle_remove_domain into try...catch with error reporting I think. No need to crash the process, if something is wrong in the removal code.